File: src/cmprlb.f
c> \file cmprlb.f

c> \brief This subroutine computes r=-Z'B(xcp-xk)-Z'g by using 
c>        wa(2m+1)=W'(xcp-x) from subroutine cauchy.
c> 
c> This subroutine computes r=-Z'B(xcp-xk)-Z'g by using 
c> wa(2m+1)=W'(xcp-x) from subroutine cauchy.
c> 
c> @param n number of parameters
c> @param m history size of Hessian approximation
c> @param x position
c> @param g gradient
c> @param ws part of L-BFGS matrix
c> @param wy part of L-BFGS matrix
c> @param sy part of L-BFGS matrix
c> @param wt part of L-BFGS matrix
c> @param z The generalized Cauchy point xcp computed by cauchy.
c>          Used here as the linearisation point: r encodes -B(z-x) - g
c>          restricted to the free variables.
c> @param r On exit r(1:nfree) contains -theta*(z(k)-x(k)) - g(k) plus the
c>          W*M^{-1}*W' correction, where k = index(i). Caller uses this as
c>          the residual for the subspace minimisation problem.
c>          When cnstnd=.false. and col>0, the shortcut path sets
c>          r(1:n) = -g(:) without using the index array.
c> @param wa Length-4m workspace shared with cauchy. On entry, the segment
c>           wa(2m+1 : 2m+2col) holds W'(z-x) (filled by cauchy). On exit
c>           wa(1 : 2col) holds M^{-1}*W'(z-x) from the bmv call.
c> @param index Permutation of (1..n): index(1..nfree) lists the indices of
c>              variables that are free at the GCP and are the active
c>              optimisation variables here.
c> @param theta Scaling factor specifying the initial Hessian B_0 = theta*I.
c> @param col Number of stored (s,y) correction pairs (0 on the first
c>            iteration; up to m thereafter).
c> @param head Index in the cyclic WS/WY buffer of the oldest stored
c>             correction. Used to walk the columns in chronological order.
c> @param nfree Number of free variables; size of the subspace problem.
c> @param cnstnd .true. if the problem has bounds; controls the shortcut
c>               path described under @param r.
c>
c> Historical note: this routine used to take an `info` output parameter
c> to forward errors from the embedded `bmv` call. Since `bmv` cannot
c> fail under LAPACK `dtrsm`, the parameter was always 0 on exit and
c> has been removed.
      subroutine cmprlb(n, m, x, g, ws, wy, sy, wt, z, r, wa, index,
     +                 theta, col, head, nfree, cnstnd)

      logical          cnstnd
      integer          n, m, col, head, nfree, index(n)
      double precision theta,
     +                 x(n), g(n), z(n), r(n), wa(4*m),
     +                 ws(n, m), wy(n, m), sy(m, m), wt(m, m)
c
c                           *  *  *
c
c     NEOS, November 1994. (Latest revision June 1996.)
c     Optimization Technology Center.
c     Argonne National Laboratory and Northwestern University.
c     Written by
c                        Ciyou Zhu
c     in collaboration with R.H. Byrd, P. Lu-Chen and J. Nocedal.
c
c
c     ************
 
      integer          i,j,k,pointr
      double precision a1,a2

      if (.not. cnstnd .and. col .gt. 0) then 
         do 26 i = 1, n
            r(i) = -g(i)
  26     continue
      else
         do 30 i = 1, nfree
            k = index(i)
            r(i) = -theta*(z(k) - x(k)) - g(k)
  30     continue
         call bmv(m,sy,wt,col,wa(2*m+1),wa(1))
         pointr = head
         do 34 j = 1, col
            a1 = wa(j)
            a2 = theta*wa(col + j)
            do 32 i = 1, nfree
               k = index(i)
               r(i) = r(i) + wy(k,pointr)*a1 + ws(k,pointr)*a2
  32        continue
            pointr = mod(pointr,m) + 1
  34     continue
      endif

      return

      end

File: docs/spec/reference_impl/core/cmprlb.py
"""cmprlb -- reduced gradient r at the Cauchy point.

Reference implementation following ``docs/spec/subroutines/cmprlb.md``.
Depends on ``core.bmv`` for the inverse-middle-matrix solve.
"""

from __future__ import annotations

import numpy as np

from .bmv import bmv


def cmprlb(
    n: int,
    m: int,
    x: np.ndarray,
    g: np.ndarray,
    ws: np.ndarray,
    wy: np.ndarray,
    sy: np.ndarray,
    wt: np.ndarray,
    z: np.ndarray,
    r: np.ndarray,
    wa: np.ndarray,
    index: np.ndarray,
    theta: float,
    col: int,
    head: int,
    nfree: int,
    cnstnd: bool,
) -> None:
    """Compute reduced gradient r at the Cauchy point.

    The F77 routine used to take an ``info`` output for forwarding bmv
    failures. Since LAPACK ``dtrsm`` cannot fail on a non-singular
    factor, that signaling is no longer needed.
    """
    # Path A: unconstrained with history -> r = -g.
    if (not cnstnd) and col > 0:
        for i in range(n):
            r[i] = -g[i]
        return

    # Path B: constrained or col == 0.
    for i in range(nfree):
        k = index[i] - 1                      # convert to 0-based
        r[i] = -theta * (z[k] - x[k]) - g[k]

    # Embedded bmv solve. Provides M^{-1} W'(z - x) in wa[1..2*col]
    # (Fortran 1-based) -> wa[0..2*col-1] in Python.
    if col > 0:
        v_in = wa[2 * m : 2 * m + 2 * col].copy()       # wa(2m+1..2m+2col)
        p_out = np.empty(2 * col, dtype=np.float64)
        bmv(m, sy, wt, col, v_in, p_out)
        wa[: 2 * col] = p_out

        pointr = head
        for j in range(1, col + 1):                       # 1-based j
            a1 = wa[j - 1]                                # wa(j)
            a2 = theta * wa[col + j - 1]                  # theta * wa(col + j)
            for i in range(nfree):
                k = index[i] - 1
                r[i] += wy[k, pointr - 1] * a1 + ws[k, pointr - 1] * a2
            pointr = (pointr % m) + 1
